src/sage/geometry/cone_critical_angles.py

```diff
@@ -0,0 +1,998 @@
+    allows our consumer to bail out immediately (knowing the optimal
+    pair!) if it turns out that the maximal angle is acute; i.e. if
+    the smallest entry of `M` is nonnegative.
+
+    INPUT:
+
+    - ``gs`` -- a linearly independent list of unit-norm generators
+      for the cone `P`
+
+    - ``hs`` -- a linearly independent list of unit-norm generators
+      for the cone `Q`
+
+    OUTPUT:
+
+    A tuple containing four elements, in order:
+
+    - The matrix `M` described in Proposition 6
+
+    - The minimal entry in the matrix `M`
+
+    - A vector in ``gs`` that achieves that minimal inner product
+      along with the next element of the tuple
+
+    - A vector in ``hs`` that achieves the minimal inner product
+      along with the previous element in the tuple
+
+    EXAMPLES::
+
+        sage: from sage.geometry.cone_critical_angles import compute_gevp_M
+        sage: P = Cone([ (1,2,0), (3,4,0) ])
+        sage: Q = Cone([ (-1,4,1), (5,-2,-1),  (-1,-1,5) ])
+        sage: gs = [g.change_ring(QQ) for g in P]
+        sage: hs = [h.change_ring(QQ) for h in Q]
+        sage: M = compute_gevp_M(gs, hs)[0]
+        sage: all( M[i][j] == gs[i].inner_product(hs[j])
+        ....:       for i in range(P.nrays())
+        ....:       for j in range(Q.nrays()) )
+        True
+
+    TESTS:
+
+    The products `(G_{I})^{T}H_{J}` correspond to
+    submatrices of the "generalized Gram matrix" `M` in Proposition
+    6. Note that SageMath does (row,column) indexing but [Or2020]_
+    does (column,row) indexing::
+
+        sage: from sage.geometry.cone_critical_angles import (
+        ....:   _random_admissible_cone,
+        ....:   compute_gevp_M,
+        ....:   gevp_licis)
+        sage: n = ZZ.random_element(1,4)
+        sage: n = ZZ.random_element(1,8) # long time
+        sage: P = _random_admissible_cone(ambient_dim=n)
+        sage: Q = _random_admissible_cone(ambient_dim=n)
+        sage: gs = [g.change_ring(QQ) for g in P]
+        sage: hs = [h.change_ring(QQ) for h in Q]
+        sage: M = compute_gevp_M(gs,hs)[0]
+        sage: f = lambda i,j: gs[i].inner_product(hs[j])
+        sage: expected_M = matrix(QQ, P.nrays(), Q.nrays(), f)
+        sage: M == expected_M
+        True
+        sage: G = matrix.column(gs)
+        sage: H = matrix.column(hs)
+        sage: def _test_indexing(I,J):
+        ....:      G_I = G[range(n), I]
+        ....:      H_J = H[range(n), J]
+        ....:      return (G_I.transpose()*H_J == M[I,J]
+        ....:              and
+        ....:              H_J.transpose()*G_I == M.transpose()[J,I])
+        sage: G_index_sets = list(gevp_licis(G))
+        sage: H_index_sets = list(gevp_licis(H))
+        sage: all( _test_indexing(I,J) for I in G_index_sets
+        ....:                          for J in H_index_sets )
+        True
+    """
+    min_u = gs[0]
+    min_v = hs[0]
+    min_ip = min_u.inner_product(min_v)
+
+    M = []
```

Comment:
```suggestion
    M = [None] * len(gs)
```
It is faster to initialize the list to the correct size if known.

Reply:
It doesn't seem to be faster in this case?

Reply:
Definitely a micro-optimization, and the benefit depends on the size of the list:
```python
sage: def test1(n):
....:     L = [None]*n
....:     for i in range(n):
....:         L[i] = i
sage: def test2(n):
....:     L = []
....:     for i in range(n):
....:         L.append(i)
sage: %timeit test1(5)
516 ns ± 14.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
sage: %timeit test2(5)
426 ns ± 9.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
sage: %timeit test1(20)
923 ns ± 25.5 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
sage: %timeit test2(20)
969 ns ± 38.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
sage: %timeit test1(1000)
34.6 µs ± 1.92 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
sage: %timeit test2(1000)
43.9 µs ± 2.18 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Reply:
Ah, all my examples were small. In that case I'd rather leave this alone. The algorithm is combinatorial in the size of `gs` and `hs`,  so you're not going be solving problems where `M` is large enough to see any benefit here.

Reply:
Okay, no problem.